File: GO-Lang/Tree/Service/BinarySearchTree.go
package service

import "fmt"

type (
	TreeNode struct {
		data  int
		right *TreeNode
		left  *TreeNode
	}
)

func InitBST(data int) *TreeNode {
	return &TreeNode{data: data}
}

func Search(root *TreeNode, value int) bool {
	if root.data == value {
		return true
	} else if root == nil {
		return false
	} else if root.data > value {
		return Search(root.left, value)
	} else {
		return Search(root.right, value)
	}
}

func InOrder(root *TreeNode) {
	if root != nil {
		InOrder(root.left)
		fmt.Println("-->", root.data)
		InOrder(root.right)
	}
}

func Insert(root *TreeNode, value int, prev *TreeNode) {
	if root == nil {
		root = &TreeNode{data: value}
		if prev.data > value {
			prev.left = root
		} else {
			prev.right = root
		}
	} else {
		if root.data > value {
			Insert(root.left, value, root)
		} else {
			Insert(root.right, value, root)
		}
	}
}

func Delete(root *TreeNode, value int) *TreeNode {
	if root == nil {
		return nil
	}
	if root.left == nil && root.right == nil {
		root = nil
		return root
	}
	if root.data > value {
		root.left = Delete(root.left, value)
	} else if root.data < value {
		root.right = Delete(root.right, value)
	} else {
		iPre := getInorderPre(root)
		root.data = iPre.data
		root.left = Delete(iPre, iPre.data)
	}
	return root
}

func getInorderPre(root *TreeNode) *TreeNode {
	root = root.left
	for root.right != nil {
		root = root.right
	}
	return root
}

File: Java/Tree/BinarySearchTree/BinarySearchTree.java
package Tree.BinarySearchTree;

import Tree.Util.TreeNode;

public class BinarySearchTree {

    public boolean search(TreeNode root, int data){
        if (root==null) return false;
        else {
            if (root.getData()==data) return true;
            else if (root.getData()>data) return search(root.getLeft(), data);
            else return search(root.getRight(), data);
        }
    }

    public void insert(TreeNode root, int data, TreeNode prev){
        if (root==null) {
            TreeNode newNode = new TreeNode(data);
            if (prev.getData()>data) prev.setLeft(newNode);
            else prev.setRight(newNode);
        }else {
            if (root.getData()==data){
                System.out.println("Duplicate value error");
                return;
            } else if (root.getData()>data) {
                insert(root.getLeft(), data, root);
            }
            else {
                insert(root.getRight(), data, root);
            }
        }
    }

    public void inOrderTraverse(TreeNode root){
        if (root!=null){
            inOrderTraverse(root.getLeft());
            System.out.println(root.getData());
            inOrderTraverse(root.getRight());
        }
    }

    public TreeNode delete(TreeNode root, int data){
        if (root == null){
            System.out.println("Tree is empty");
            return null;
        } else if (root.getRight()==null && root.getLeft()==null) {
            root = null;
            return null;
        } else if (root.getData()>data) {
            root.setLeft(delete(root.getLeft(), data));
        } else if (root.getData()<data) {
            root.setRight(delete(root.getRight(), data));
        }else {
            TreeNode iPre = getInOrderPre(root);
            root.setData(iPre.getData());
            root.setLeft(delete(iPre, iPre.getData()));
        }
        return root;
    }

    private TreeNode getInOrderPre(TreeNode root) {
        root = root.getLeft();
        while (root.getRight()!=null) root = root.getRight();
        return root;
    }
}
